File: Leetcode 0017 Letter Combinations of a Phone Number.py
## 2025/07/03
## backtracking
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        dic = {
            '2':"abc",
            '3':"def",
            '4':"ghi",
            '5':"jkl",
            '6':"mno",
            '7':"pqrs",
            '8':"tuv",
            '9':"wxyz",
        }
        def backtrack(idx, comb):
            if idx == len(digits):
                res.append(comb[:])
                return 
            for letter in dic[digits[idx]]:
                backtrack(idx+1, comb+letter)
        res =[]
        backtrack(0,"")
        return res
## easy way
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        l = len(digits)
        if l == 0:
            return []
        ans = []
        dic = {
            '2':"abc",
            '3':"def",
            '4':"ghi",
            '5':"jkl",
            '6':"mno",
            '7':"pqrs",
            '8':"tuv",
            '9':"wxyz",
        }
        if l == 1:
            return [i for i in dic[digits]]

        elif l == 2:
            for b in dic[digits[1]]:
                for a in dic[digits[0]]:
                    ans.append(f"{a}{b}")
            return ans
        elif l == 3:
            for c in dic[digits[2]]:
                for b in dic[digits[1]]:
                    for a in dic[digits[0]]:
                        ans.append(f"{a}{b}{c}")
            return ans
        elif l == 4:
            for d in dic[digits[3]]:
                for c in dic[digits[2]]:
                    for b in dic[digits[1]]:
                        for a in dic[digits[0]]:
                            ans.append(f"{a}{b}{c}{d}")
            return ans

File: Leetcode 0017 Letter Combinations of a Phone Number.go
// 2025/07/03

func letterCombinations(digits string) []string {
    ans := make([]string,0)
    l := len(digits)
    if l == 0{
        return ans
    }
    dic := make(map[string][]string)
    dic["2"] = append(dic["2"], []string{"a","b","c"}...)
    dic["3"] = append(dic["3"], []string{"d","e","f"}...)
    dic["4"] = append(dic["4"], []string{"g","h","i"}...)
    dic["5"] = append(dic["5"], []string{"j","k","l"}...)
    dic["6"] = append(dic["6"], []string{"m","n","o"}...)
    dic["7"] = append(dic["7"], []string{"p","q","r","s"}...)
    dic["8"] = append(dic["8"], []string{"t","u","v"}...)
    dic["9"] = append(dic["9"], []string{"w","x","y","z"}...)
    if l == 1{
        for _, a := range dic[string(digits[0])]{
            ans = append(ans, a)
        }
    }   else if l == 2 {
            for _, b := range dic[string(digits[1])]{
                for _, a := range dic[string(digits[0])]{
                    ans = append(ans, a+b)
            }
        }
    }   else if l == 3 {
            for _, c := range dic[string(digits[2])]{
                for _, b := range dic[string(digits[1])]{
                    for _, a := range dic[string(digits[0])]{
                        ans = append(ans, a+b+c)
                    }
                }
            }
        } else if l == 4 {
            for _, d := range dic[string(digits[3])]{
                for _, c := range dic[string(digits[2])]{
                    for _, b := range dic[string(digits[1])]{
                        for _, a := range dic[string(digits[0])]{
                            ans = append(ans, a+b+c+d)
                        }
                    }
                }
            }
        }


    return ans // 暫時 return nil，避免編譯錯誤
}
